sway/desktop/output.c

```diff
@@ -984,14 +1077,17 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 
 static void handle_present(struct wl_listener *listener, void *data) {
 	struct sway_output *output = wl_container_of(listener, output, present);
-	struct wlr_output_event_present *output_event = data;
+	struct wlr_output_event_present *event = data;
 
-	if (!output->enabled || !output_event->presented) {
+	if (!output->enabled || !event->presented) {
 		return;
 	}
 
-	output->last_presentation = *output_event->when;
-	output->refresh_nsec = output_event->refresh;
+	// Send frame done to all visible surfaces
+	struct send_frame_done_data frame_done_data = {0};
+	clock_gettime(CLOCK_MONOTONIC, &frame_done_data.when);
+	frame_done_data.msec_until_refresh = get_msec_until_refresh(event);
+	send_frame_done(output, &frame_done_data);
```

Comment:
It seems strange for the compositor's `handle_present()` to also need to call `get_msec_until_refresh()` in order to determine the frame callback delay. `get_msec_until_refresh()` uses `output->max_render_time`, which is specific to the `timed_frame_scheduler` implemented here. For a generic frame scheduler, we don't seem to expose how long this delay should be, and assuming `output->max_render_time` seems wrong.

- Should frame callback timers be set by the scheduler in its present function instead?
- If it makes more sense to continue setting the callbacks on the compositor side, maybe this is better addressed in the frame scheduler MR?

I know this is on the back-burner, and things have changed a bit in the `wlr_output` events since the last rebase, so feel free to ignore if this no longer makes sense.

Reply:
> It seems strange for the compositor's handle_present() to also need to call get_msec_until_refresh() in order to determine the frame callback delay

Why do you think so? `handle_present` notifies us of when presentation occurred some time after the fact and is not representative of the actual time of presentation. We need to do a bit of math to figure out how much time is actually left at this point, using the reported time of presentation and estimated time from there till the next frame.

> `get_msec_until_refresh()` uses `output->max_render_time`

Hmm? It doesn't in the PR as is. `output->max_render_time` is applied by `timed_frame_scheduler` itself to offset render from the value reported by `get_msec_until_refresh`.

> For a generic frame scheduler, we don't seem to expose how long this delay should be

I believe the intent was to later implement a "fancy" frame scheduler that dynamically decides on a suitable delay, which would more or less deprecate max_render_time.

Reply:
> Hmm? It doesn't in the PR as is. output->max_render_time is applied by timed_frame_scheduler itself to offset render from the value reported by get_msec_until_refresh.

Whoops, you're right, `output->max_render_time` is used by `send_frame_done_iterator` not in `get_msec_until_refresh`. That got mixed up at some point between this implementation, master, and my own patch :sweat_smile: 

Same point though, `output->max_render_time` is used outside of the frame scheduler.

> I believe the intent was to later implement a "fancy" frame scheduler that dynamically decides on a suitable delay, which would more or less deprecate max_render_time.

We're on the same page here. There are technically 2 delays here a frame scheduler implementation might want to configure separately. We're using `max_render_time` for both right now (ignoring `view->max_render_time`), but there is:
- A) the delay between present and the frame event (which triggers rendering for the compositor)
- B) the delay between present and the frame callbacks being called (which triggers rendering for clients)

(A) is already being handled entirely within this frame scheduler. My question is about how (B) should be handled by the frame scheduler, since in this implementation it's not.

Reply:
Having messed around with implementing a frame scheduler for a bit, I now realize that it doesn't make sense for (B) to be implemented in `wlr_frame_scheduler`. Even though it is very similar to (A), it is probably more appropriate to handle per-surface, rather than per-output.

So I guess this PR doesn't need to address that, sorry for the noise.